torch/_functorch/aot_autograd.py

```diff
@@ -256,309 +539,145 @@ def from_fun(t):
-        #     will not allow you to mutate it (This original came from arbitrary user code where the user might want to mutate)
-        # (2) There's no reason to compile views anyway. We can just regenerate the view of the input off to the side,
-        #
-        # Another interesting case is when you have both mutation and aliasing:
-        # def f(x):
-        #     x.mul_(2)
-        #     return x.view(-1)
-        #
-        # You could imagine that this output is now *safe* to compile and return in the autograd.Function,
-        # because after functionalization runs, it will technically not alias an input:
-        # def f_functionalized(x):
-        #     x_updated = x.mul(2)
-        #     return x_updated, x_updated.view(-1)
-        #
-        # However, this is still wrong: we can't return x_updated.view(-1) to the user. We are on the hook to return:
-        # def traced_forward(x):
-        #     x_updated = x.mul(2)
-        #     return x_updated
-        #
-        # def compiled_fn(x)
-        #     x_updated = traced_forward(x)
-        #     x.copy_(x_updated)
-        #     return x.view(-1)
-        #
-        # Why can't we return x_updated.view(-1) to the user?
-        # It can have different metadata from x.view(-1)! Specifically, the input x could be a non-memory-dense tensor,
-        # But the intermediate created by our graph, x_updated, will always be memory-dense.
-        def filter_and_record_aliased_outs(outputs):
-            # NOTE: this dict will clobber keys if we have multiple inputs that alias.
-            # Let's say inpA and inpB alias, and the user generated an output using out = inpA.view(...)
-            # For now, since we're not handling the case with multiple _base's sharing a storage,
-            # it is actually fine to arbitrarily pick which input to regenerate the aliased output from.
-            # e.g. out_new = inpB.as_strided(out.size(), out.stride(), out.storage_offset())
+                mutation_type = MutationType.none
+
+            input_info.append(InputAliasInfo(mutation_type=mutation_type))
+
+        # If a function involves creating a tensor, and returning a view of it, such that its _base is the intermediiate,
+        # We need to make sure our graph returns the _base as a graph output, and we manually recreate the view
+        # to return to the user. Why? The backend compiler is free to (incorrectly) not set requires_grad
+        # on the base tensor, but we are obligated to properly set requires-gradness on the real output.
+
+        num_mutated_inps = len([x for x in input_info if x.mutation_type != MutationType.none])
+        inp_storage_refs = {
+            StorageWeakRef(inpt.storage()): idx
+            for idx, inpt in enumerate(flat_f_args) if isinstance(inpt, torch.Tensor)
+        }
+        inp_tensor_ids = {id(inpt) for inpt in flat_f_args if isinstance(inpt, torch.Tensor)}
+        out_tensor_ids = {id(o): i for i, o in enumerate(flat_f_outs)}
```

Comment:
This two variables are weirdly asymmetric

Reply:
In this case, we map from the out tensor's id to its idx so that we can regenerate an alias off of that tensor. So if two output tensors have the same id, it shouldn't matter which tensor we use to regenerate the alias.

I also added a quick dummy test:
```
         def f(a):
             out = torch.mul(a, 3)
             out_view = out.view(-1)
             return out, out_view, out
```

Reply:
they're asymmetric because for we only need the input tensor id's to know if any outputs **are** inputs. We need the output tensor ids to know if an output's `._base` attributes **are** existing outputs, so we can generate the alias off of the correct output. I can try to clarify a bit in the comments